File: fibonacci_last_digit_optimised.cpp
#include <iostream>
#include<cassert>
using namespace std;

int fast_last(int n) {
        
        
    int arr[60];
    
    arr[0] = 0;
    arr[1] = 1;

    int previous = 0;
    int current  = 1;

    for (int i = 2; i < 60; ++i) {
        int tmp_previous = previous;
        previous = current;
        current = tmp_previous + current;
        arr[i] = current%10;
    }
    int res_index = n%60;
    int result = arr[res_index];
    return result;
}

/*int get_fibonacci_last_digit_naive(int n) {
    if (n <= 1)
        return n;

    int previous = 0;
    int current  = 1;

    for (int i = 0; i < n - 1; ++i) {
        int tmp_previous = previous;
        previous = current;
        current = tmp_previous + current;
    }

    return current % 10;
}
*/


int main() {
    int n;
    std::cin >> n;
    int c = fast_last(n);
    //int d = get_fibonacci_last_digit_naive(n);
    //assert(fast_last(n)==get_fibonacci_last_digit_naive(n));
    //cout<<d<<" - Non optimised"<<"\n";
    std::cout << c;
    
}

File: fibonacci_last_digit_optimised.py
# cook your dish here
import sys


def f_last(n):
    last_digit=[0,1,1]
    t1=0
    t2=1 
    nextnumber=1
    for _ in range(3,60):
        t1 = t2
        t2=nextnumber
        nextnumber = t1+t2
        last_digit.append(nextnumber%10)
        
    res_index = n%60
    return last_digit[res_index]



'''def period(b):
    for i in range(1, b*b+1):
        if f(i)%b == 0 and f(i+1)%b == 1:
            return(i)
    function to find the periods of particular base last digit repetition 
    
   '''         
            
            

def get_fibonacci_last_digit_naive(n):
    if n <= 1:
        return n

    previous = 0
    current  = 1

    for _ in range(n - 1):
        previous, current = current, previous + current

    return current % 10

if __name__ == '__main__':
    input = sys.stdin.read()
    n = int(input)
    #assert(get_fibonacci_last_digit_naive(n)!=f_last(n))
    print(f_last(n))
    #print(get_fibonacci_last_digit_naive(n))
    
        
